File: 398/398599/398599_01.c
/*
 * A398599 -- direct exhaustive reference implementation.
 *
 * Count permutations p of {1,...,n} satisfying
 *
 *        |p(i+1) - p(i)| != s,   1 <= i < n.
 *
 * The default is s=3.  This program follows the definition directly.  It
 * fills the permutation from left to right and tries every unused value.
 * A branch is rejected exactly when the newly appended value differs from
 * the preceding value by s.  It uses no inclusion-exclusion formula, tile
 * partition, generating function, or memoization.
 *
 * For n>s, runtime is O(n!) in the worst case and working memory is O(n).
 * The input guard n<=30 is an implementation limit, not a practical speed
 * claim: s=3, n=30 is far beyond the reach of exhaustive enumeration.
 * GMP and a safely flushed unsigned-long leaf counter make the final count
 * exact without integer overflow.  If n<=s, the condition is vacuous and
 * the program returns n! directly.
 *
 * --upto prints a(0),...,a(MAX_N) as one comma-separated line.  No b-file
 * is created.
 *
 * Build (Homebrew GMP on Apple silicon):
 *
 *   clang -O3 -std=c11 -Wall -Wextra -Wpedantic \
 *     -I/opt/homebrew/include -L/opt/homebrew/lib \
 *     398599_01.c -lgmp -o 398599_01
 *
 * Usage:
 *
 *   ./398599_01                     # s=3, print a(0)..a(10)
 *   ./398599_01 --upto 10
 *   ./398599_01 --term 10
 *   ./398599_01 --s 4 --upto 10
 *   ./398599_01 --check             # check s=3 through n=10
 *   ./398599_01 --check 11
 */

#define _POSIX_C_SOURCE 200809L

#include <errno.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <gmp.h>

#define DEFAULT_S 3
#define DEFAULT_MAX_N 10
#define DEFAULT_CHECK_N 10
#define MAX_SUPPORTED_N 30
#define MAX_SUPPORTED_S 128
#define KNOWN_MAX_N 12

/* A398599, n=0..12. */
static const char *const known_terms[KNOWN_MAX_N + 1] = {
    "1",
    "1",
    "2",
    "6",
    "12",
    "48",
    "240",
    "1392",
    "9936",
    "82416",
    "771888",
    "8081808",
    "93273936"
};

typedef struct {
    int n;
    int s;
    bool used[MAX_SUPPORTED_N];
    mpz_ptr total;
    unsigned long pending;
} SearchContext;

typedef enum {
    MODE_UPTO,
    MODE_TERM
} OutputMode;

static void die(const char *message)
{
    fprintf(stderr, "error: %s\n", message);
    exit(EXIT_FAILURE);
}

static double monotonic_seconds(void)
{
    struct timespec now;
    if (clock_gettime(CLOCK_MONOTONIC, &now) != 0) {
        die("clock_gettime failed");
    }
    return (double)now.tv_sec + (double)now.tv_nsec / 1000000000.0;
}

static int parse_integer(const char *text, const char *label,
                         int minimum, int maximum)
{
    char *end = NULL;
    errno = 0;
    long value = strtol(text, &end, 10);
    if (errno != 0 || end == text || *end != '\0' ||
        value < minimum || value > maximum) {
        fprintf(stderr, "error: %s must be in %d..%d: %s\n",
                label, minimum, maximum, text);
        exit(EXIT_FAILURE);
    }
    return (int)value;
}

static int parse_n(const char *text, const char *label)
{
    return parse_integer(text, label, 0, MAX_SUPPORTED_N);
}

static int parse_s(const char *text)
{
    return parse_integer(text, "S", 1, MAX_SUPPORTED_S);
}

static void flush_pending(SearchContext *context)
{
    if (context->pending != 0UL) {
        mpz_add_ui(context->total, context->total, context->pending);
        context->pending = 0UL;
    }
}

static void record_permutation(SearchContext *context)
{
    if (context->pending == ULONG_MAX) {
        flush_pending(context);
    }
    ++context->pending;
}

/*
 * Directly fill position 0,1,...,n-1.  Values are represented by
 * 0,...,n-1 instead of 1,...,n; this does not change their differences.
 */
static void enumerate(SearchContext *context, int position, int last_value)
{
    if (position == context->n) {
        record_permutation(context);
        return;
    }

    for (int value = 0; value < context->n; ++value) {
        if (context->used[value]) {
            continue;
        }
        if (position > 0) {
            int difference = value - last_value;
            if (difference == context->s || difference == -context->s) {
                continue;
            }
        }

        context->used[value] = true;
        enumerate(context, position + 1, value);
        context->used[value] = false;
    }
}

static void verify_known_value(const mpz_t value, int n, int s)
{
    if (s != DEFAULT_S || n > KNOWN_MAX_N) {
        return;
    }

    mpz_t expected;
    mpz_init(expected);
    if (mpz_set_str(expected, known_terms[n], 10) != 0) {
        mpz_clear(expected);
        die("invalid built-in decimal value");
    }
    if (mpz_cmp(value, expected) != 0) {
        gmp_fprintf(stderr,
                    "error: A398599 mismatch at n=%d: got %Zd, "
                    "expected %s\n",
                    n, value, known_terms[n]);
        mpz_clear(expected);
        exit(EXIT_FAILURE);
    }
    mpz_clear(expected);
}

static void compute_value(mpz_t result, int n, int s, bool verbose)
{
    double started = monotonic_seconds();

    if (n <= s) {
        mpz_fac_ui(result, (unsigned long)n);
    } else {
        SearchContext context;
        context.n = n;
        context.s = s;
        memset(context.used, 0, sizeof(context.used));
        context.total = result;
        context.pending = 0UL;

        mpz_set_ui(result, 0UL);
        enumerate(&context, 0, 0);
        flush_pending(&context);
    }

    verify_known_value(result, n, s);
    if (verbose) {
        fprintf(stderr,
                "398599_01: s=%d, n=%d, direct exhaustive DFS, %.3f s\n",
                s, n, monotonic_seconds() - started);
    }
}

static void print_value(const mpz_t value)
{
    if (mpz_out_str(stdout, 10, value) == 0U) {
        die("could not write result");
    }
}

static int print_sequence(int maximum_n, int s)
{
    mpz_t value;
    mpz_init(value);
    for (int n = 0; n <= maximum_n; ++n) {
        compute_value(value, n, s, false);
        if (n != 0) {
            fputs(", ", stdout);
        }
        print_value(value);
    }
    putchar('\n');
    mpz_clear(value);
    return EXIT_SUCCESS;
}

static int check_implementation(int maximum_n)
{
    if (maximum_n > KNOWN_MAX_N) {
        fprintf(stderr, "error: CHECK_N must be in 0..%d: %d\n",
                KNOWN_MAX_N, maximum_n);
        return EXIT_FAILURE;
    }

    mpz_t value;
    mpz_init(value);
    for (int n = 0; n <= maximum_n; ++n) {
        compute_value(value, n, DEFAULT_S, false);
    }
    mpz_clear(value);

    printf("ok: direct enumeration agrees with A398599 for "
           "s=%d, n=0..%d\n",
           DEFAULT_S, maximum_n);
    return EXIT_SUCCESS;
}

static void usage(const char *program)
{
    fprintf(stderr,
            "usage: %s [MAX_N]\n"
            "       %s --term N\n"
            "       %s --upto MAX_N\n"
            "       %s --s S [MAX_N]\n"
            "       %s --s S --term N\n"
            "       %s --s S --upto MAX_N\n"
            "       %s --check [CHECK_N]\n"
            "\n"
            "S defaults to %d and may be 1..%d.\n"
            "MAX_N defaults to %d and may be 0..%d.\n"
            "--upto prints a(0)..a(MAX_N), comma-separated.\n"
            "No b-file is created.\n",
            program, program, program, program, program, program, program,
            DEFAULT_S, MAX_SUPPORTED_S, DEFAULT_MAX_N, MAX_SUPPORTED_N);
}

int main(int argc, char **argv)
{
    const char *program = argv[0];
    if (argc == 2 &&
        (strcmp(argv[1], "--help") == 0 || strcmp(argv[1], "-h") == 0)) {
        usage(program);
        return EXIT_SUCCESS;
    }

    if (argc >= 2 && strcmp(argv[1], "--check") == 0) {
        if (argc > 3) {
            usage(program);
            return EXIT_FAILURE;
        }
        int maximum_n = argc == 3 ?
            parse_integer(argv[2], "CHECK_N", 0, KNOWN_MAX_N) :
            DEFAULT_CHECK_N;
        return check_implementation(maximum_n);
    }

    int s = DEFAULT_S;
    int n = -1;
    bool have_s = false;
    OutputMode mode = MODE_UPTO;

    for (int argument = 1; argument < argc; ++argument) {
        const char *text = argv[argument];
        if (strcmp(text, "--s") == 0) {
            if (have_s || argument + 1 >= argc) {
                usage(program);
                return EXIT_FAILURE;
            }
            s = parse_s(argv[++argument]);
            have_s = true;
        } else if (strcmp(text, "--term") == 0 ||
                   strcmp(text, "--upto") == 0) {
            if (n >= 0 || argument + 1 >= argc) {
                usage(program);
                return EXIT_FAILURE;
            }
            mode = strcmp(text, "--term") == 0 ? MODE_TERM : MODE_UPTO;
            n = parse_n(argv[++argument],
                        mode == MODE_TERM ? "N" : "MAX_N");
        } else {
            if (text[0] == '-' || n >= 0) {
                usage(program);
                return EXIT_FAILURE;
            }
            n = parse_n(text, "MAX_N");
            mode = MODE_UPTO;
        }
    }

    if (n < 0) {
        n = DEFAULT_MAX_N;
    }
    if (mode == MODE_UPTO) {
        return print_sequence(n, s);
    }

    mpz_t value;
    mpz_init(value);
    compute_value(value, n, s, true);
    printf("%d ", n);
    print_value(value);
    putchar('\n');
    mpz_clear(value);
    return EXIT_SUCCESS;
}

File: 398/398599/398599_01.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Exact computation of b_{1,s}(n):
#
#   #{p in S_n : |p(i+1)-p(i)| != s for every 1 <= i < n}.
#
# This evaluates the Spahn--Zeilberger tiling-partition formula
#
#   b_{r,s}(n) = sum_alpha C_alpha(n,r) C_alpha(n,s)
#                  * (-1)^(number_of_tiles-n)
#                  * product_j a_j! * 2^(a_2+a_3+...).
#
# Here r=1, while s is selected on the command line.  Ruby Integer gives
# exact arbitrary-precision arithmetic.  No b-file is created; --upto
# prints a(0)..a(MAX_N) as one comma-separated line.
#
# Reference:
#   G. Spahn and D. Zeilberger,
#   "Counting Permutations Where The Difference Between Entries Located
#    r Places Apart Can never be s",
#   https://doi.org/10.54550/ECA2023V3S2R10
#
# Usage:
#   ./398599_01.rb
#   ./398599_01.rb --upto 30
#   ./398599_01.rb --s 2 --upto 20
#   ./398599_01.rb --term 20
#   ./398599_01.rb --check

module B1S
  DEFAULT_S = 3
  DEFAULT_MAX_N = 20
  MAX_SUPPORTED_N = 30
  MAX_SUPPORTED_S = 128
  BRUTE_CHECK_MAX_N = 11
  GENERAL_CHECK_MAX_S = 8

  # For n<=30 there are at most fifteen non-singleton tiles, so five bits
  # per histogram count are sufficient.  Packed histograms can be combined
  # by ordinary Integer addition without inter-field carries.
  HISTOGRAM_BITS = 5
  HISTOGRAM_MASK = (1 << HISTOGRAM_BITS) - 1

  class CalculationError < StandardError; end

  # OEIS A002464 = b_{1,1}(n), n=0..20.
  KNOWN_B11 = %w[
    1 1 0 0 2 14 90 646 5242 47622 479306 5296790 63779034
    831283558 11661506218 175203184374 2806878055610
    47767457130566 860568917787402 16362838542699862
    327460573946510746
  ].map!(&:to_i).freeze

  module_function

  def parse_integer(text, label, minimum, maximum)
    value = Integer(text, 10)
    unless value.between?(minimum, maximum)
      raise CalculationError,
            "#{label} must be in #{minimum}..#{maximum}: #{text}"
    end
    value
  rescue ArgumentError
    raise CalculationError,
          "#{label} must be in #{minimum}..#{maximum}: #{text}"
  end

  def parse_n(text, label)
    parse_integer(text, label, 0, MAX_SUPPORTED_N)
  end

  def parse_s(text)
    parse_integer(text, 'S', 1, MAX_SUPPORTED_S)
  end

  def binomial(n, k)
    return 0 if k.negative? || k > n

    k = [k, n - k].min
    value = 1
    1.upto(k) { |i| value = value * (n - k + i) / i }
    value
  end

  # Enumerate all possible non-singleton tile histograms for one path.
  # edge_count is sum_l l*h_l and component_count is sum_l h_l.
  def generate_local_recursive(path_size, length_limit, length_index,
                               edge_count, component_count,
                               component_orders, packed_histogram, output)
    if length_index == length_limit
      gap_choices = binomial(path_size - edge_count, component_count)
      output << [packed_histogram, gap_choices * component_orders]
      return
    end

    length = length_index + 1
    remaining_vertices = path_size - edge_count - component_count
    maximum = remaining_vertices / (length + 1)
    0.upto(maximum) do |count|
      next_histogram =
        packed_histogram | (count << (HISTOGRAM_BITS * length_index))
      next_orders = component_orders *
                    binomial(component_count + count, count)
      generate_local_recursive(
        path_size, length_limit, length_index + 1,
        edge_count + length * count, component_count + count,
        next_orders, next_histogram, output
      )
    end
  end

  def generate_local_patterns(path_size)
    raise CalculationError, 'invalid residue path size' if path_size < 1

    output = []
    generate_local_recursive(
      path_size, path_size - 1, 0, 0, 0, 1, 0, output
    )
    output.freeze
  end

  # Return H -> C_H(n,step), together with convolution statistics.
  def tiling_coefficients(n, step)
    path_sizes = 1.upto(step).map { |k| (n - k) / step + 1 }
    current = { 0 => 1 }
    cache = {}
    convolution_pairs = 0

    path_sizes.each do |path_size|
      patterns = cache[path_size] ||= generate_local_patterns(path_size)
      following = Hash.new(0)
      current.each do |histogram, coefficient|
        patterns.each do |local_histogram, local_coefficient|
          following[histogram + local_histogram] +=
            coefficient * local_coefficient
          convolution_pairs += 1
        end
      end
      current = following
    end

    coefficient_sum = current.values.sum
    expected_sum = 1 << (n - step)
    unless coefficient_sum == expected_sum
      raise CalculationError,
            "step-#{step} tiling coefficients do not sum to 2^(n-step)"
    end
    [current, convolution_pairs]
  end

  def factorial_table(n)
    factorial = Array.new(n + 1, 1)
    1.upto(n) { |i| factorial[i] = factorial[i - 1] * i }
    factorial
  end

  def compute_formula(n, s)
    unless n.between?(0, MAX_SUPPORTED_N) &&
           s.between?(1, MAX_SUPPORTED_S)
      raise CalculationError, 'unsupported n or s'
    end

    started = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    factorial = factorial_table(n)
    if n <= s
      seconds = Process.clock_gettime(Process::CLOCK_MONOTONIC) - started
      return [
        factorial[n],
        {
          input_histograms: 1,
          output_histograms: 1,
          convolution_pairs: 0,
          seconds: seconds
        }
      ]
    end

    input, input_pairs = tiling_coefficients(n, 1)
    output, output_pairs = tiling_coefficients(n, s)
    result = 0

    # The step-1 path may contain tiles of every size up to n.
    input.each do |packed_histogram, input_coefficient|
      output_coefficient = output[packed_histogram]
      next if output_coefficient.nil?

      histogram = packed_histogram
      edge_count = 0
      component_count = 0
      histogram_factorial = 1
      0.upto(n - 2) do |length_index|
        count = histogram & HISTOGRAM_MASK
        unless count.zero?
          edge_count += (length_index + 1) * count
          component_count += count
          histogram_factorial *= factorial[count] if count > 1
        end
        histogram >>= HISTOGRAM_BITS
      end

      singleton_count = n - edge_count - component_count
      if singleton_count.negative?
        raise CalculationError, 'invalid common tile histogram'
      end

      term = input_coefficient * output_coefficient
      term *= factorial[singleton_count] * histogram_factorial
      term <<= component_count # two orientations per non-singleton tile
      result += edge_count.even? ? term : -term
    end

    unless result.between?(0, factorial[n])
      raise CalculationError, 'formula result is outside 0..n!'
    end

    seconds = Process.clock_gettime(Process::CLOCK_MONOTONIC) - started
    [
      result,
      {
        input_histograms: input.length,
        output_histograms: output.length,
        convolution_pairs: input_pairs + output_pairs,
        seconds: seconds
      }
    ]
  end

  def verify_known_value(value, n, s)
    return unless s == 1 && n < KNOWN_B11.length
    return if value == KNOWN_B11[n]

    raise CalculationError,
          "A002464 mismatch at n=#{n}: " \
          "got #{value}, expected #{KNOWN_B11[n]}"
  end

  def compute_value(n, s, verbose:)
    value, stats = compute_formula(n, s)
    verify_known_value(value, n, s)
    if verbose
      warn format(
        '%s: r=1, s=%d, n=%d, input histograms=%d, ' \
        'output histograms=%d, convolution pairs=%d, %.3f s',
        File.basename($PROGRAM_NAME), s, n,
        stats[:input_histograms], stats[:output_histograms],
        stats[:convolution_pairs], stats[:seconds]
      )
    end
    value
  end

  # Independent subset DP: build the permutation from left to right and
  # remember only the used values and the last value.
  def brute_recursive(n, s, full_mask, memo, used, last_value)
    return 1 if used == full_mask

    index = used * (n + 1) + last_value
    cached = memo[index]
    return cached unless cached.nil?

    total = 0
    available = full_mask ^ used
    until available.zero?
      bit = available & -available
      available ^= bit
      value = bit.bit_length - 1
      next if last_value < n && (value - last_value).abs == s

      total += brute_recursive(n, s, full_mask, memo,
                               used | bit, value)
    end
    memo[index] = total
  end

  def brute_count(n, s)
    unless n.between?(0, BRUTE_CHECK_MAX_N)
      raise CalculationError,
            'independent check n is outside its supported range'
    end

    full_mask = (1 << n) - 1
    memo = Array.new((1 << n) * (n + 1))
    brute_recursive(n, s, full_mask, memo, 0, n)
  end

  def check_implementation(maximum_n)
    0.upto(maximum_n) do |n|
      value, = compute_formula(n, 1)
      verify_known_value(value, n, 1)
    end

    brute_maximum = [maximum_n, BRUTE_CHECK_MAX_N].min
    1.upto(GENERAL_CHECK_MAX_S) do |s|
      0.upto(brute_maximum) do |n|
        value, = compute_formula(n, s)
        independent = brute_count(n, s)
        next if value == independent

        raise CalculationError,
              "independent DP mismatch at s=#{s}, n=#{n}: " \
              "formula=#{value}, DP=#{independent}"
      end
    end

    known_maximum = [maximum_n, KNOWN_B11.length - 1].min
    puts "ok: A002464=b_{1,1}(n) verified through n=#{known_maximum}; " \
         "formula evaluated through n=#{maximum_n}; " \
         "independent DP for s=1..#{GENERAL_CHECK_MAX_S} " \
         "through n=#{brute_maximum}"
    0
  end

  def print_sequence(maximum_n, s)
    values = 0.upto(maximum_n).map do |n|
      compute_value(n, s, verbose: false)
    end
    puts values.join(', ')
    0
  end

  def usage(program)
    warn <<~USAGE
      usage: #{program} [MAX_N]
             #{program} --term N
             #{program} --upto MAX_N
             #{program} --s S [MAX_N]
             #{program} --s S --term N
             #{program} --s S --upto MAX_N
             #{program} --check [MAX_N]

      S defaults to #{DEFAULT_S} and may be 1..#{MAX_SUPPORTED_S}.
      MAX_N defaults to #{DEFAULT_MAX_N} and may be 0..#{MAX_SUPPORTED_N}.
      --upto prints b_{1,S}(0)..b_{1,S}(MAX_N), comma-separated.
      No b-file is created.
    USAGE
  end

  def main(arguments)
    program = File.basename($PROGRAM_NAME)
    if arguments.length == 1 && %w[--help -h].include?(arguments[0])
      usage(program)
      return 0
    end
    if arguments.first == '--check'
      if arguments.length > 2
        usage(program)
        return 1
      end
      maximum_n = arguments.length == 2 ?
        parse_n(arguments[1], 'MAX_N') : DEFAULT_MAX_N
      return check_implementation(maximum_n)
    end

    mode = :upto
    s = DEFAULT_S
    n = nil
    have_s = false
    argument = 0
    while argument < arguments.length
      text = arguments[argument]
      case text
      when '--s'
        if have_s || argument + 1 >= arguments.length
          usage(program)
          return 1
        end
        argument += 1
        s = parse_s(arguments[argument])
        have_s = true
      when '--term', '--upto'
        if !n.nil? || argument + 1 >= arguments.length
          usage(program)
          return 1
        end
        mode = text == '--term' ? :term : :upto
        argument += 1
        n = parse_n(arguments[argument], mode == :term ? 'N' : 'MAX_N')
      else
        if text.start_with?('-') || !n.nil?
          usage(program)
          return 1
        end
        n = parse_n(text, 'MAX_N')
        mode = :upto
      end
      argument += 1
    end

    n ||= DEFAULT_MAX_N
    if mode == :term
      value = compute_value(n, s, verbose: true)
      puts "#{n} #{value}"
      0
    else
      print_sequence(n, s)
    end
  end
end

begin
  exit B1S.main(ARGV)
rescue B1S::CalculationError => error
  warn "error: #{error.message}"
  exit 1
end
